### PR TITLE
Allow publishers to tag stats announcements in whitehall

### PR DESCRIPTION
**Pull request by klssmith**
Publishers are now able to tag statistical announcements to the new taxonomy.
As with tagging documents, this is only enabled for statistical announcements belonging 
to certain organisations. Tagging statistical announcements is not mandatory.

[Trello card](https://trello.com/c/HSMA7TyX/87-allow-publishers-to-tag-stats-announcements-in-whitehall)

### Screenshots
screen shot 2017-06-05 at 10 53 58

screen shot 2017-06-05 at 10 54 40

### CODE REVIEW

File: app/controllers/admin/statistics_announcement_tags_controller.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+  def edit
+    @govuk_taxonomy = Taxonomy::GovukTaxonomy.new
+    @tag_form = TaxonomyTagForm.load(@statistics_announcement.content_id)
+  end
+
+  def update
+    @tag_form = TaxonomyTagForm.new(
+      content_id: @statistics_announcement.content_id,
+      selected_taxons: params["taxonomy_tag_form"].fetch("taxons", []).reject(&:blank?),
```
Comment (whoojemaflip) [inline app/controllers/admin/statistics_announcement_tags_controller.rb]: this is sufficiently complex to warrant it's own method.

File: app/controllers/admin/statistics_announcements_controller.rb [hunk 2]
```diff
@@ -8,6 +8,10 @@ def index
     @statistics_announcements = @filter.statistics_announcements
   end
 
+  def show
+    fetch_expanded_links if @statistics_announcement.can_be_tagged_to_taxonomy?
```
Comment (whoojemaflip) [inline app/controllers/admin/statistics_announcements_controller.rb]: might it be more explicit to set the `@expanded_links` variable here?

File: app/models/statistics_announcement.rb [hunk 3]
```diff
@@ -207,8 +207,24 @@ def requires_redirect?
...
+  end
 
 private
+
+  def organisations_in_tagging_beta?(org_content_ids)
+    return false if org_content_ids.empty?
+
+    organisations_in_tagging_beta = Whitehall.organisations_in_tagging_beta
+
+    org_content_ids.any? do |id|
```
Comment (whoojemaflip) [inline app/models/statistics_announcement.rb]: This runs at  O(n<sup>2</sup>) so maybe very slow if either of the arrays is sizeable. 

Could it work with an intersection instead? `(organisations_in_tagging_beta & org_content_ids).present?`

**Comment by klssmith** [comment]
@whoojemaflip Thanks - I've addressed your comments

File: app/views/admin/statistics_announcements/show.html.erb [hunk 4]
```diff
@@ -69,6 +69,33 @@
         <%= link_to "Cancel statistics release", [:cancel, :admin, @statistics_announcement], class: "btn btn-default pull-right" %>
       <% end %>
     </div>
+
+    <% if @statistics_announcement.can_be_tagged_to_taxonomy? %>
+      <section class="taxonomy-topics" id="topic-new-taxonomy">
```
Comment (whoojemaflip) [inline app/views/admin/statistics_announcements/show.html.erb]: this is quite a complex bit of view code. Could it be made into a partial? How much does it have in common with the `edition` version?

**Pull request merged**